### PR TITLE
Split `ONNXParser` from `MarabouONNXNetwork` in preparation for replacing with C++ implementation

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
This commit starts to lay the ground work for #727 by extracting the parsing components of `MarabouONNXNetwork` and `MarabouNetwork` into two new classes `ONNXParser` and `InputQueryBuilder`. These two new classes will in a future PR be directly implemented by the existing `ONNXParser.h` and `NetworkParser.h` on the C++ side (the latter to be renamed to `InputQueryBuilder` at later stage).

In particular, I've also changed `ONNXParser.py` to be a pure operation, simply adding constraints to an `InputQueryBuilder`, rather than storing the constraints internally. This will greatly ease the later implementation of support for queries over multiple networks.

I've striven to maximise backwards compatibility as much as possible, and I haven't broken any example code. The only two unavoidable changes were as follows:
  1.  `MarabouONNXNetwork` no longer exposes the fields `madeGraphEquations`, `varMap`, `constantMap`, `shapeMap`. 
This I think is a distinct improvement as these are really internal implementation details of the parser, and have nothing to do with the constraints generated.
  2. `MarabouONNXNetwork` no longer has a `shallowCopy` method. Instead of calling this method,
    there's a new parameter `preserveExistingConstraints` in the method `readONNX` to
    `True` which has the same effect and is much more semantically meaningful I think!

### CODE REVIEW

**Review by wu-haoze — CHANGES_REQUESTED** [review]

File: maraboupy/MarabouNetwork.py [hunk 1]
```diff
@@ -584,35 +284,6 @@ def findError(self, inputValues, options=None, filename="evaluateWithMarabou.log
...
-                equivalence = False
-        for inputvars1, inputvars2 in zip(self.inputVars, network.inputVars):
-            if (inputvars1.flatten() != inputvars2.flatten()).any():
-                equivalence = False
-        for outputVars1, outputVars2 in zip(self.outputVars, network.outputVars):
-            if (outputVars1.flatten() != outputVars1.flatten()).any():
-                equivalence = False
-        return equivalence
-
     def addConstraint(self, constraint: VarConstraint):
```
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply maraboupy/MarabouNetwork.py]: Yup good idea! Done :+1:

**Comment by wu-haoze** [comment]
Another general thought is that we could just create a MarabouCore.InputQuery object at the initialization of InputQueryBuilder(), and directly encode bounds, equations, ReLUs into the InputQuery, as the methods in the InputQueryBuilder() class are called.

This is probably beyond the scope of this PR, but would be necessary when we replace in the C++ ONNX parser later on. Because the C++ ONNX parser will create a marabouCore.InputQuery object. 

@MatthewDaggitt Let me know what you think! I'm happy to do this refactoring.

File: maraboupy/parsers/InputQueryBuilder.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            isProperty (bool): If true, this constraint can be removed later by clearProperty() method
+        """
+        assert len(vars)==len(coeffs)
+        e = MarabouUtils.Equation(MarabouCore.Equation.LE)
+        for i in range(len(vars)):
+            e.addAddend(coeffs[i], vars[i])
+        e.setScalar(scalar)
+        self.addEquation(e, isProperty)
+
+    def getMarabouQuery(self):
```
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply maraboupy/parsers/InputQueryBuilder.py]: Have done :+1:

File: maraboupy/parsers/InputQueryBuilder.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+All rights reserved. See the file COPYING in the top-level source
+directory for licensing information.
+'''
+
+from maraboupy import MarabouCore
+from maraboupy import MarabouUtils
+from maraboupy.MarabouPythonic import *
+from abc import ABC
+
+class InputQueryBuilder(ABC):
```
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply maraboupy/parsers/InputQueryBuilder.py:22]: Yes, that's exactly the idea. It's eventually going to be way more flexible in the way you can build queries. 

If it's okay with you, I would like to leave updating the documentation until we have discussed the final design in the meeting next week?

[the remark this answers is not in the document]
Comment (wu-haoze) [reply maraboupy/parsers/InputQueryBuilder.py:22]: Awesome, these all sound great to me!

File: maraboupy/MarabouNetwork.py [hunk 4]
```diff
@@ -17,12 +17,12 @@
 '''
 
 from maraboupy import MarabouCore
-from maraboupy import MarabouUtils
 from maraboupy.MarabouPythonic import *
```
Comment (wu-haoze) [inline maraboupy/MarabouNetwork.py:20]: This import could be removed.

`from maraboupy.MarabouPythonic import *`

**Review by wu-haoze — APPROVED** [review]

**Comment by MatthewDaggitt** [comment]
> Another general thought is that we could just create a MarabouCore.InputQuery object at the initialization of InputQueryBuilder(), and directly encode bounds, equations, ReLUs into the InputQuery, as the methods in the InputQueryBuilder() class are called.
>
> This is probably beyond the scope of this PR, but would be necessary when we replace in the C++ ONNX parser later on. Because the C++ ONNX parser will create a marabouCore.InputQuery object.

Yes, eventually I think we should be unifying `InputQuery` and `InputQueryBuilder`!

File: maraboupy/MarabouNetwork.py (hunk 4 above)
Comment (MatthewDaggitt) [reply maraboupy/MarabouNetwork.py]: Fixed.

**Comment by wu-haoze** [comment]
> > Another general thought is that we could just create a MarabouCore.InputQuery object at the initialization of InputQueryBuilder(), and directly encode bounds, equations, ReLUs into the InputQuery, as the methods in the InputQueryBuilder() class are called.
> > This is probably beyond the scope of this PR, but would be necessary when we replace in the C++ ONNX parser later on. Because the C++ ONNX parser will create a marabouCore.InputQuery object.
> 
> Yes, eventually I think we should be unifying `InputQuery` and `InputQueryBuilder`!

I think there is a motivation to have an extra layer (InputQueryBuilder) because of the user-friendly python specific methods for adding constraints. But the layer should be made thiner than what it is now..

**Comment by MatthewDaggitt** [comment]
> I think there is a motivation to have an extra layer (InputQueryBuilder) because of the user-friendly python specific methods for adding constraints. But the layer should be made thiner than what it is now..

Yes, okay. I'm definitely planning to unify the C++ and the Python at the `InputQueryBuilder` layer. After that we can look at it and reassess the situation :+1:

**Pull request merged**